Get index of first non ascii char

pgovind opened the pull request:
I think this is ready and can be reviewed now. 

Implements GetIndexOfFirstNonAsciiChar from #35034

Updated Perf:
```
| Faster                                                                 | base/diff | Base Median (ns) | Diff Median (ns) | Modality|
| ---------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| -------- |
| System.Text.Experimental.Perf.IsNormalized_GetIndexOfFirstNonAsciiChar |      1.40 |        111784.29 |         79682.34 |         |
```

Commit:
GetIndexOfFirstNonAsciiChar

Commit:
Temp commit to change branches

Commit:
Bug fix

Commit:
revert commenting out test case

eiriktsarpalis commented on src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs:
```diff
@@ -747,6 +787,18 @@ private static unsafe nuint GetIndexOfFirstNonAsciiChar_Sse2(char* pBuffer, nuin
                     }
+                    else if (AdvSimd.Arm64.IsSupported)
+                    {
+                        currentMask = Unicode.Utf16Utility.GetNonAsciiBytes(AdvSimd.AddSaturate(combinedVector, asciiMaskForAddSaturate).AsByte(), bitmask);
```
Minor nit: The `goto FoundNonAsciiDataInFirstOrSecondVector;` statement is replicated across many branches, which could make the logic harder to modify longer-term. Consider factoring SIMD logic into a predicate method, e.g. `bool ContainNonAsciiDataInFirstOrSecondVector()` and then consume like so:

```csharp
if (ContainNonAsciiDataInFirstOrSecondVector(..args..))
{
    goto FoundNonAsciiDataInFirstOrSecondVector;
}
```

kunalspathak commented:
Do you mind sending a PR to `dotnet/performance` to add `IsNormalized_GetIndexOfFirstNonAsciiChar ()` benchmark? We need it to track .NET 3.1 vs. .NET 5 performance improvements in this area.

kunalspathak commented on src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs:
```diff
@@ -647,8 +647,12 @@ private static unsafe nuint GetIndexOfFirstNonAsciiChar_Sse2(char* pBuffer, nuin
+
+            Vector128<byte> bitmask = BitConverter.IsLittleEndian ?
+                Vector128.Create(0x80402010_08040201).AsByte() :
+                Vector128.Create(0x01020408_10204080).AsByte();
```
I think this is also true for other PRs that Carlos sent, but can you remind me why we use `bitMask` for `!IsLittleEndian` if it is not supported? Is the idea that when we start supporting, the code will just work?

pgovind replied:
Yup, that's the idea.

kunalspathak commented on src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs:
```diff
@@ -673,13 +677,35 @@ private static unsafe nuint GetIndexOfFirstNonAsciiChar_Sse2(char* pBuffer, nuin
+            }
+            else if (AdvSimd.Arm64.IsSupported)
+            {
+                currentMask = Unicode.Utf16Utility.GetNonAsciiBytes(AdvSimd.AddSaturate(firstVector, asciiMaskForAddSaturate).AsByte(), bitmask);
```
You should check out excellent feedback from @TamarChristinaArm  in https://github.com/dotnet/runtime/pull/39050#issuecomment-662458778 and https://github.com/dotnet/runtime/pull/39050#issuecomment-668699538 about optimizing this code. I think you should follow the same because in the end, all you are doing this calling `BitOperations.TrailingZeroCount()` on it.

TamarChristinaArm replied:
Agreed with splitting it off. In this case it looks like this code only cares about the index of the first non-zero element. So we can do even better here.

If you use a mask `0x0f00` and `BIC` you can get a much more efficient sequence (see https://github.com/ARM-software/optimized-routines/blob/224cb5f67b71757b99fe1e10b5a437c17a1d733c/string/aarch64/strlen.S#L164)

essentially

```
sshr	v1.16b, v1.16b, 7
bic	v1.8h, 0x0f, lsl 8
umaxp	v1.16b, v1.16b, v1.16b
fmov	x1, d1
clz	x0, x1
```

as the sequence to get the first element that has the `msb` set. Of course for the cases where it's just doing `if (currentMask != 0)` you just need a `maxp` and `fmov` as I explained in the other post.

TamarChristinaArm replied:
this is for little endian btw, for big-endian you need a slight variantion. though this can be avoided loading with `LD1` instead of `LDR` (which I think is what `LoadVector128` does here).

kunalspathak reviewed (changes requested):
You should not map the SSE2 logic for AdvSimd. `AdvSimd` performance can be better and doesn't have to go through  `AddSaturate()` logic.

pgovind commented:
Updated Perf: 

```
Updated Perf:
```
| Faster                                                                 | base/diff | Base Median (ns) | Diff Median (ns) | Modality|
| ---------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| -------- |
| System.Text.Experimental.Perf.IsNormalized_GetIndexOfFirstNonAsciiChar |      1.40 |        111784.29 |         79682.34 |         |
```

cc @kunalspathak

pgovind commented:
Filed https://github.com/dotnet/runtime/issues/40805 to follow up on Tamar's suggestions here. Considering that we're already seeing decent perf improvements here, how do folks feel about merging this PR now and investigating the suggestions in a future PR?

kunalspathak commented:
Couple of things:
- There was some feedback given in https://github.com/dotnet/runtime/pull/39103#discussion_r455552959 that I don't see we implemented. It was also brought up by Tamar, the key point is we should not mimic the SSE2 logic of `AddSaturate()` because ARM64 has better instructions to do similar operation. I agree that it will take some time to make it most efficient implementation, but we should at least think about ways to do the operation without `AddSaturate()`, etc. Currently, we are doing more than needed operations like 3 `AddPairwise` which might show improvement on WSL2 but might have degraded performance on different processors.
- If I see the code of `IsNormalized_GetIndexOfFirstNonAsciiChar()` in https://github.com/dotnet/performance/pull/1445/files#diff-2d837d38e3d94ab0d7f80e232693857eR90 , I see that you are testing it on ascii data instead of non-ascii data. Was that intentional? Are the above results from the ascii version as seen in the PR in performance repo?

With that, I am not sure if we should rush this in for RC1.

jeffhandley commented:
@pgovind I'm going to close this PR since this work is on hold. When we get back around to this work, we can create a fresh PR.
